File: 0001~0500/0074. Search a 2D Matrix/main.cpp
/** 
 * Source: https://leetcode.com/problems/search-a-2d-matrix/
 * Date: 2022/7/14
 * Skill: Binary Search
 * Runtime: 
 * Memory Usage: 
 * Time complexity: 
 * Space complexity: 
 * Constraints: 
 *     
 */

#include<iostream>
#include<vector>
#include<string>
#include<cmath>   
//#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int mid = matrix.size()*matrix[0].size()/2, start = 0, end = matrix.size()*matrix[0].size()-1, 
        col = matrix[0].size(), row = matrix.size();
        while (start <= mid && mid <= end) {
            if (matrix[mid/col][mid%col] == target)
                return true;
            else if (matrix[mid/col][mid%col] > target) {
                end = mid-1;
                mid = (mid+start)/2;
            } else {
                start = mid+1;
                mid = (mid+end+1)/2;
            }
        }
        return false;
    }
};

int main() {
    Solution s;
    
}

static const auto io_sync_off = []() {
    //* turn off sync
    std::ios::sync_with_stdio(false);
    //* untie in/out streams
    std::cin.tie(nullptr);
    return nullptr;
}();

File: 0001~0500/0074. Search a 2D Matrix/main.py
"""
Source: https://rb.gy/fubp4
Date: 2023/8/7
Skill:
Ref:
Runtime: 52 ms, faster than 84.30%
Memory Usage: 16.79 MB, less than 90.55%
Time complexity:
Space complexity:
Constraints:

"""

import math, sys
from typing import List, Optional
from collections import defaultdict, Counter, deque
from heapq import heapify, heappush, heappop, nsmallest
import heapq
import functools
from bisect import bisect_left, bisect_right


class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        if matrix[0][0] > target or matrix[-1][-1] < target:
            return False
        lo, hi = 0, len(matrix) - 1
        while lo < hi:
            mid = (lo + hi + 1) // 2
            if matrix[mid][0] > target:
                hi = mid - 1
            else:
                lo = mid
        left, right = 0, len(matrix[0]) - 1
        while left < right:
            mid = (left + right) // 2
            if matrix[lo][mid] == target:
                return True
            elif matrix[lo][mid] < target:
                left = mid + 1
            else:
                right = mid

        return False if matrix[lo][left] != target else True


if __name__ == "__main__":
    s = Solution()
